# RibosePreferenceAnalysis
Preference analysis for ribose-seq data

This script is used for preference analysis of rNMP incorporation data generated from _ribose-seq_ protocol and Ribose-Map software. The neighbor of incorporated rNMP keeps certain preference in a specific species. This is because the surrounding dNMPs could affect the probability of rNMP misincorporation by replicative polymerases. This software is designed to reveal this preference.

## Dependency

Except Python3 standard libraries. Those packages are needed:

- Matplotlib

- Seaborn

- Scipy

## Usage

1. Use __dinuc_count.py__ to calculate background frequencies for the reference genome (__FASTA__ file)

   ```bash
   ./dinuc_count.py <ref genome> -o <background frequency>
   ```

   Dinucleotide background frequency is calculated by default. Available parameters are:

   1. __-d D__  Distance between dinucleotide pair, default = 1
   
   2. __-s__  Count only one strand
   
   3. __--mono__  Count mono nucleotide instead
   
   4. __--trinuc__  Count trinucleotide instead
   
2.  Count different patterns of rNMP incorporation from __BED__ file generated by Ribose-Map using __dinuc_cnt_ribose.py__.

   ```bash
   ./dinuc_cnt_ribose.py <ref genome> <BED> -o <rNMP incorporation raw>
   ```

   By default only dinucleotide frequency is counted. Available parameters:

   1. __-f__  Use fourth column of bed file as frequency. Otherwise each row of BED file is considered as a rNMP
   2. __-m__  Also count mononucleotide frequency
   3. __-d__  Also count dinucleotide frequency
   4. __--dist DIST [DIST ...]__  Distance between rNMP and its dNMP neighbor
   5. __-t__  Also count trinucleotide frequency
   
4. Get the data of desired chromosome.

   ```bash
   ./get_chrom.py <infile1> <infile2> ... -s <chromosome1> <chromsome2> ... -o <file desired>
   ```

   Available parameters:

   1. __-v__  Select non-matching chromosomes 
   2. __-a__  Append to original file 
   3. __--name NAME__  Name for the output line, default = input file name

4. Normalization

   ```bash
   ./normalize.py <file desired> <background frequency> -o <normalized file>
   ```

   Available parameters:

   1. __--group_len {0,4,16}__  Number of rows of which the sum is 1. If 0 is selected, the sum of all rows will be 1. default = 0.
   2. __--name NAME__  Name of chromosome in background frequency used for normalization, default = saccer
   
5. Draw preference heatmaps.

   ```bash
   ./draw_ribose.py <normalized file> -o <output basename>
   ```

   A heatmap in __PNG__ format is generated by default. Available parameters:

   1. __-b B__  Select background file. If a file is selected, the background percentage is added to labels.
   2. __--nr__  Input file is NR type dinucleotide frequency. Of which the second position is rNMP.
   3. __--mono__  Input file is mononucleotide frequency.
   4. __--tri TRI__  Input file is trinucleotide frequency, with rNMP incorporated at TRI position
   5. __--background_chrom BACKGROUND_CHROM__  Chromosome name of background file, default = chrM
   6. __--legend_group {0,4,16}__  Number of lables of which the sum is 1. If 0 is selected, the sum of all labels will be 1. default = 0.
   7. __--no_annot__  Hide percentage annotation in each cell.
   8. __--cmax CMAX__  Maximum value in color scale. Any preferency beyond that will show as the maximum color.
   
   ## License
   
   This software is under GNU GPL v3.0 license
   
   ## Contact
   
   If you have any question, please contact me at [pxu64@gatech.edu](mailto:pxu64@gatech.edu).
   
   Please give me a star if you like it!

